### ISSUE TITLE
Adopt zerosearch-node for docs search; retire minsearch

### ISSUE THREAD

**Issue by alexeygrigorev** [DataTalksClub/dataops#85]
# Adopt `zerosearch-node` for docs search; retire `minsearch`

Status: blocked (until #84 closes)
Tags: `migration`, `backend`, `P2`
Depends on: #84 (and the `SearchIndex` contract defined in #83)
Blocks: #87 (final integration), #88

Part of the single-TypeScript-backend consolidation (#83). See
[`_docs/TARGET_ARCHITECTURE.md`](../blob/main/_docs/TARGET_ARCHITECTURE.md) →
"zerosearch-node / DataOps adoption".

## Scope

Build a TypeScript search module that implements the **`SearchIndex`** contract
from #83, backed by `zerosearch-node` (`~/git/zerosearch-node`, #84). Stage it
under `work-engine/src/docs/search/` (promoted to `backend/src/docs/` at #88).
Provide a Node index builder that reads `content/` markdown and writes the
portable `json-1` index, replacing the Python
`build_search_index.py` / `docs_index.py` / `search_handler.py` behavior.

This issue delivers the **search module + Node index builder + updated smoke
fixtures + parity evidence vs `minsearch`**. The live HTTP `/search` endpoint is
wired by the content port (#87), which consumes this module — keep that wiring
out of this issue but make the module ready for it. Do not leave two live index
paths; the Python search path is removed at the #87/#88 cutover.

Field config carries over **exactly** from
`lambda-functions/src/lambda_functions/docs_index.py`:

- text fields & boosts: `title`(4), `summary`(3), `description`(3),
  `purpose`(3), `headings`(2), `body`(1), `tags`, `systems`
- keyword fields: `path`, `id`, `domain`, `doc_type`

## Acceptance Criteria

- [ ] `SearchIndex` TS implementation under `work-engine/src/docs/search/`,
      backed by `zerosearch-node`, matching the #83 contract
      (`build`/`load`, `search(query, { filter, boost, numResults })`).
- [ ] Field config matches `docs_index.py` exactly (text fields + boosts and
      keyword fields listed above), including how each field is derived from
      frontmatter/markdown.
- [ ] A Node index builder rebuilds the index over `content/` and writes the
      portable `json-1` format; the documented refresh command replaces
      `python -m lambda_functions.build_search_index`.
- [ ] Search **smoke fixtures updated for TF-IDF → BM25-lite**: recall expected
      on par, ordering may shift. Capture a `minsearch` top-k baseline, diff it,
      and record the ranking change with reasons in the issue/PR notes.
- [ ] Keyword filtering (`domain`, `doc_type`, `path`, `id`) does exact-match
      and per-field boosts apply, matching today's behavior.
- [ ] `minsearch` + numpy/scipy/scikit-learn/pandas are removed from the docs
      **search** path; the deploy-package check that confirms their full absence
      is finalized at the #88 packaging cutover.
- [ ] Index-build location decision recorded (see Dependencies).

### Scenario: smoke query relevance
Given: an index built over `content/`
When: the existing smoke query set runs through `SearchIndex.search`
Then: the expected docs appear in top-k (recall on par with `minsearch`);
updated fixtures reflect the new BM25-lite ordering.

### Scenario: keyword filter + boost
Given: docs across multiple `domain`/`doc_type` values
When: searching with a `filter` on `domain` and a `boost` on `title`
Then: only matching-domain docs return and title matches rank higher.

## Dependencies

- **`json-1` index format / cross-language interop (decision needed).**
  Python `zerosearch` serializes with `marshal` (interpreter/platform-pinned,
  not language-neutral), so `zerosearch-node` defined a portable **`json-1`**
  format (see `~/git/zerosearch-node/FORMAT.md` and the finding on #84). For
  DataOps the cleanest path is to **move index build fully to Node** so no
  Python reader is needed. **Native bidirectional interop** (a Python
  `zerosearch` that reads/writes `json-1`) is **not done** and would require
  modifying the **published PyPI** `zerosearch` library — do not assume it
  exists. Decision before this issue closes:
  - (a) **preferred** — build the index in Node (zerosearch-node writes
    `json-1`), removing Python from search entirely; **or**
  - (b) add a `json-1` reader/writer to published Python `zerosearch` (separate
    task, needs the go-ahead to touch a published lib).
  This issue assumes (a). Record the choice in the issue.
- Keep a captured `minsearch` top-k snapshot as the recall baseline for the
  fixture update.

## Out of Scope

- Live HTTP `/search` endpoint wiring and search-refresh-after-save (→ #87).
- SOP engine (#86); deleting Python files / package + Lambda cutover (#88).
- Frontend search UI changes.
- Modifying the published Python `zerosearch` library (only flagged here).

## Tests

- `npm --prefix work-engine test` — new `SearchIndex`/builder module tests.
- `zerosearch-node` BM25-lite parity is already covered by #84.
- Document and run the Node index-rebuild command over `content/`.

**Comment by alexeygrigorev** [comment]
## READY FOR TEST — #85 (zerosearch-node docs search)

Implemented the docs `SearchIndex` on `zerosearch-node` plus a Node index builder over `content/`, with updated BM25-lite smoke fixtures and a minsearch parity baseline. Scope stayed within the owned paths; no edits to `sop/`, `contentApi.ts`, `githubStore.ts`, or `router.ts`.

### Consumption decision (zerosearch-node)
`zerosearch-node` is not published to npm and lives at `~/git/zerosearch-node` (outside the repo). Chosen path: **vendor it in-repo and reference it via a relative `file:` dependency** (`work-engine/vendor/zerosearch-node`, `"zerosearch-node": "file:vendor/zerosearch-node"`).
- An absolute `file:~/git/...` path or a git dep breaks `npm ci` on CI runners; vendoring the prebuilt artifact makes `npm ci` + Lambda packaging deterministic and network-free. It is zero-dependency, so nothing transitive is pulled in.
- The work-engine is CommonJS (and deploys as a CJS Lambda); upstream `zerosearch-node` ships **ESM-only** `dist/`, which a CJS `require()` cannot load (`ERR_PACKAGE_PATH_NOT_EXPORTED`). So the vendored copy is **recompiled from upstream `src/` to CommonJS** (a clean two-file recompile). Verified it loads under plain `node` (not just tsx), i.e. Lambda-viable.
- The committed `dist/` is force-tracked via a `.gitignore` exception (`!vendor/zerosearch-node/dist/`).
- Proper npm publish + switching to a semver range is a **deploy follow-up** at the #88 packaging cutover (recorded in `work-engine/vendor/README.md`).

### Index-build location decision
Took option **(a)**: the index is built fully in Node and written as the portable **`json-1`** format (no Python reader needed). Verified header: `{"magic":"zerosearch","format":"json-1",...}`. Default output `work-engine/.tmp/search/docs.index` (gitignored); the deployed/runtime location is finalized by the content port (#87).

### Implemented vs deferred
- `searchIndex.ts` — `ZeroSearchIndex` / `createSearchIndex` / `loadSearchIndex` on `zerosearch-node`, reusing `DOCS_SEARCH_CONFIG` unchanged. `search(query, { filter, boost, numResults })`: config boosts overlaid by per-query boosts, keyword `filter` passthrough, results mapped to `_score`. `save`/`load` use `json-1`.
- `search/extract.ts` — port of `docs_index.iter_docs`: frontmatter via `gray-matter`, markdown cleaning (code/image/link/HTML/marker stripping), `extractHeadings`, `scalarFrontmatterValue`, `inferDomain`/`inferDocType`, and `generatedId` (from `doc_registry._generated_id`). `iterContentDocs()` walks `content/`.
- `scripts/build-search-index.ts` — Node builder replacing `python -m lambda_functions.build_search_index`. Refresh command: `npm --prefix work-engine run build:search-index` (indexed **357** docs).
- **Deferred (TODO #86):** SOP-structure-aware body extraction (`doc_to_search_text` -> `sop_parse.parse`). The SOP engine port landed under `src/docs/sop/` but is wired into the live content path only at #87, so `docToSearchText` currently indexes the cleaned raw markdown body and exposes a single `structuredText` hook as the marked integration point. Recall-preserving (SOP step text is present in the raw body).
- **Out of scope (unchanged):** live `/search` wiring (#87); deleting Python `minsearch`/numpy/scipy/sklearn/pandas from the package (#88 cutover).

### Fixture changes / minsearch -> BM25-lite diff
New fixtures in `tests/docs-search.test.ts` reflect BM25-lite ordering. Captured a minsearch top-5 baseline over the same `content/` and compared:

| query | top-1 | recall@5 overlap |
|---|---|---|
| podcast intake | SAME | 3/5 |
| newsletter sponsor | SAME | 3/5 |
| course certificate | SAME | 4/5 |
| youtube upload | SAME | 4/5 |

Top-1 is identical for every query; ordering shifts within the relevant set and the non-overlapping hits are still topically on-target (other podcast/newsletter/course/youtube docs). This matches the documented "recall on par, ordering may shift" expectation.

### Acceptance criteria
- [x] `SearchIndex` TS impl under docs, backed by zerosearch-node, matching the contract.
- [x] Field config matches `docs_index.py` exactly (text+boosts, keyword), incl. derivation from frontmatter/markdown (config-parity + extraction tests).
- [x] Node index builder rebuilds over `content/` -> `json-1`; documented refresh command.
- [x] Smoke fixtures updated for TF-IDF -> BM25-lite; baseline captured + diff recorded above.
- [x] Keyword filtering (`domain`/`doc_type`/`path`/`id`) exact-match + per-field boosts (filter+boost tests).
- [~] minsearch + numpy/scipy/sklearn/pandas removed from search path — the new TS path is in place; **Python deletion + deploy-package check is the #88 cutover** (out of scope here, per the issue).
- [x] Index-build location decision recorded (option (a), above).

### Verification
- `npm run typecheck:work-engine` — green.
- `npm run test:work-engine` — **999 pass / 0 fail** (incl. new `docs-search.test.ts`: 11 tests).
- `npm run build:work-engine` — compiles; emits `dist/docs/search*`.
- Runtime `require()` of emitted CJS under plain `node` — loads + searches OK (Lambda-relevant).
- `build:search-index` over `content/` — 357 docs, `json-1` artifact.

### Commits (worktree `worktree-agent-a7ab067733695fe49`, not merged)
- `b324b90` Vendor zerosearch-node (CJS) + gray-matter dep.
- `53df916` Implement ZeroSearchIndex + Node index builder + tests.

Changed/added files (all absolute under the worktree):
- `work-engine/src/docs/searchIndex.ts`
- `work-engine/src/docs/search/extract.ts`
- `work-engine/scripts/build-search-index.ts`
- `work-engine/tests/docs-search.test.ts`
- `work-engine/src/docs/index.ts` (barrel), `work-engine/package.json`, `work-engine/.gitignore`
- `work-engine/vendor/zerosearch-node/**`, `work-engine/vendor/README.md`

**Comment by alexeygrigorev** [comment]
Merged to `main` (`827e25f`). `SearchIndex` implemented on `zerosearch-node` (vendored as CJS at `work-engine/vendor/zerosearch-node`, loads under plain node); Node index builder over `content/` (357 docs → `json-1`); BM25-lite smoke fixtures with captured minsearch baseline (top-1 identical for all 4 queries, recall on par, ordering shifts as expected). typecheck + 999/999 tests green (11 new in `docs-search.test.ts`).

Deferred per scope: live `/search` HTTP wiring → #87; `#86` structured SOP extraction plugged via the `structuredText` hook in `extract.ts`; deletion of Python minsearch/numpy/scipy/sklearn/pandas + deploy-package check → #88. Vendoring→npm-publish is a #88 deploy follow-up.

**Comment by alexeygrigorev** [comment]
## On-Call: WorkEngine SAM deploy fix (Refs #85)

The `dataops-v1` stack is now deploying green. The push that fixed the work-engine e2e flake (harden-signal-wait-retry, #90/#91/#92) reached deploy for the first time and exposed a pre-existing packaging defect in `zerosearch-node`.

### What was failing
`Deploy DataOps v1 stack` → `sam deploy`:
```
Error: Unable to upload artifact WorkEngineFunction referenced by CodeUri
[Errno 2] No such file or directory: '.../WorkEngineFunction/node_modules/zerosearch-node'
```
This was masked before because every prior push died at the e2e gate (`task-detail-names.spec.js`) before deploy could run.

### Root cause
`build-WorkEngineFunction` in the root `Makefile` copied `work-engine/package.json` into the SAM artifact but not the vendored `work-engine/vendor/` dir. Since `zerosearch-node` is a `file:vendor/zerosearch-node` dependency, npm resolves it to a symlink `node_modules/zerosearch-node -> ../work-engine/vendor/zerosearch-node`, which was dangling in the artifact (target never copied). SAM upload only packages real files, so the broken symlink caused the upload failure.

### Fix (commit 7b8ff3f)
In `build-WorkEngineFunction`: copy `work-engine/vendor` into the artifact so the symlink target exists, then replace the symlink with the real directory so SAM uploads actual files. Verified the built artifact loads (`require('zerosearch-node')` + handler module both OK), and `sam build` produces a real, resolvable `node_modules/zerosearch-node`.

### CI/CD result
Run 28470614672 — green.
- Check DataOps v1 app: 4m52s (docs 121 passed, browser 3 passed, work-engine e2e 320 passed)
- Deploy DataOps v1 app: 2m6s — `Successfully created/updated stack - dataops-v1 in eu-west-1`, seed + all 3 smoke tests passed (docs full app, workflow templates, private work engine Lambda)

### Note
Proper publishing of `zerosearch-node` to npm and switching off the `file:` vendor dependency remains the deploy follow-up tracked here under the #88 packaging cutover.

**Comment by alexeygrigorev** [comment]
Adopted: work-engine/src/docs/searchIndex.ts implements the SearchIndex contract on top of zerosearch-node (commit 53df916), with a build-search-index.ts script and full test suite (docs-search.test.ts, 189 lines). The index builder runs over content/ and writes the portable json-1 format. Confirmed working in production. Closing.
